### DISCUSSION TITLE
What's the reasoning behind the updates to calculation and orthogonalization for bitangent?

### DISCUSSION THREAD

**Discussion by johannesugb** [assimp/assimp#5512] [Q&A]
Namely this commit: https://github.com/assimp/assimp/commit/ad766cb738fbc899efe64593ff9ec2fd9c0a70a1
by @youkeyao in February 2022

I think these cause problems in one of our applications (basically it leads to flipped normals), therefore I'd like to know why the changes were made.

**Comment by youkeyao** [comment]
I think this is right according to https://learnopengl.com/Advanced-Lighting/Normal-Mapping. It does work in my case, whereas the previous didn't.

**Reply by johannesugb** [reply]
Okay, I've got some answers now. 

The biggest problem with these changes is: The new code orthogonalizes the bitangent with the tangent by using Gram–Schmidt which is generally **wrong**.
The following is the offending line:
```cpp
aiVector3D localBitangent = bitangent - meshNorm[p] * (bitangent * meshNorm[p]) - localTangent * (bitangent * localTangent);
```
specifically the last/added part: `- localTangent * (bitangent * localTangent);`.

The reason why it is generally wrong is because uv parameterization for texture-mapped models is not orthogonal in most cases. See Figure 1 for an example. This has also been noted in another discussion here: https://github.com/assimp/assimp/discussions/4857#discussioncomment-4596315.

![image](https://github.com/assimp/assimp/assets/536330/661f07b0-7984-4111-8af1-7b3fedf4334e)
_Figure 1:_ An example of uv mapping onto a 3D model, showing that u and v directions are, in general, non-orthogonal.

The error introduced by using an orthogonal tangent space is, in most cases, probably pretty small because texture mapping strives to map uv coordinates uniformly, I guess. But as Figure 2 shows, the error can get quite severe.

![image](https://github.com/assimp/assimp/assets/536330/9d5bb99e-7e88-4491-a334-74899a7919c4)
_Figure 2:_ Acting as if the tangent space was orthogonal introduces quite a large error for the "skewed" case. The error is as big as the difference between the black (correct) and the red (incorrect, result of using orthogonal tangent space) arrows.

Orthogonalization is _always_ an optimization, because then computations might get cheaper---because then it is no longer required to compute the inverse of a TBN matrix, because for orthogonal matrices, the inverse==the transpose.      
ASSIMP shall not do this kind of optimization unconditionally. In fact, it shall not do it at all in my opinion. The decision of if and when (e.g. on the CPU-side or in shaders) shall be left to the application.

**First conclusion: The orthogonalization of bitangent w.r.t. the tangent must be reverted.**

The other change that was introduced by the commit actually looked good, namely flipping the handedness of the bitangents, as compared in the following code listing:
```cpp
// Good old code:                                        // New code from ad766cb:
bitangent.x = (w.x * sx - v.x * tx) * dirCorrection;     bitangent.x = (- w.x * sx + v.x * tx) * dirCorrection;
bitangent.y = (w.y * sx - v.y * tx) * dirCorrection;     bitangent.y = (- w.y * sx + v.y * tx) * dirCorrection;
bitangent.z = (w.z * sx - v.z * tx) * dirCorrection;     bitangent.z = (- w.z * sx + v.z * tx) * dirCorrection;
```
It is also in line with [Eric Lengyel -Mathematics for 3D Game Programming & Computer Graphics, Third Edition](https://www.amazon.com/dp/1435458869/?tag=terathon-20), page 182.

However, we have the following problem: Texture coordinates (or textures) generally need to be flipped vertically in order to produce the correct rendering results as illustrated in Figure 3. Flipping textures only in one direction (namely, vertically) basically corresponds to changing the handedness. The old bitangent code **compensated for that** => One could argue that this is just the ASSIMP way of calculating bitangents, as it has always been.

![image](https://github.com/assimp/assimp/assets/536330/5169b0d1-4427-496e-828f-06de5390f6c8)
_Figure 3:_ Rendering result of the Sponza 3D model with raw/non-flipped textures and UV coordinates (left => incorrect) and with UV coordinates flipped (right => correct). (Of course, _either_ UV coordinates _or_ textures must be flipped to get the desired effect.)

Now, one could argue that the application shall compensate for the flipping, but I would strongly suggest to not impose such kinds of responsibilities on the applications when ASSIMP didn't impose it before February 2022/commit https://github.com/assimp/assimp/commit/ad766cb738fbc899efe64593ff9ec2fd9c0a70a1. Especially since others struggle with the new bitangents code as well, like in Q&A #4857 .

**Second conclusion:** Also revert flipping of the handedness. => Go back to good old ASSIMP implementation.

**Reply by youkeyao** [reply]
Maybe you're right. But I have the same problem as #3191 when it goes to the old code. I'll check again to see if there are other problems.

**Reply by johannesugb** [reply]
Could you try to apply the change of handedness and/or the orthogonalization in shaders?

E.g. like this:

```glsl
layout (location = 2) in vec3 aNormal;
layout (location = 3) in vec3 aTangent;
layout (location = 4) in vec3 aBitangent;

// Re-orthogonalizes the first vector w.r.t. the second vector (Gram-Schmidt process)
vec3 re_orthogonalize(vec3 first, vec3 second)
{
	return normalize(first - dot(first, second) * second);
}

...

vec3 bitangent = re_orthogonalize(aBitangent, aTangent);
```

Other possible solutions could be to check the sign of the bitangent in shaders and flip it (depending on your concrete implementation, e.g.:

```glsl
vec3 bitangent = cross(aNormal, aTangent);
// if texture is mirrored, flip bitangent:
if (dot(bitangent, aBitangent) < 0)
{
	bitangent = -bitangent ;
}
```

Does any of these approaches help to solve your problems in #3191 ?

**Reply by youkeyao** [reply]
Thanks, my problem is solved. But I'm still wondering if it's necessary to calculate different bitangent according to different handedness since it's already set during import by `aiProcess_ConvertToLeftHanded`. People might get confused because the default model handedness is right-handed but the tangent space is left-handed. Just my suggestion.

**Reply by johannesugb** [reply]
Well, `aiProcess_ConvertToLeftHanded` cannot be the way to go because that also turns everything into clockwise face order. But `aiProcess_FlipUVs` could maybe be an option. Anyways, I'm in favor of keeping the code as it was before because then it does not break compatibility to older applications based on ASSIMP. And the option to flip the UVs is still there.
